File: c11_char_manip/src/string_manipulation.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include "string_manipulation.h"

void reverse_string(char *str) {
    int n = strlen(str);
    for (int i = 0; i < n / 2; i++) {
        char temp = str[i];
        str[i] = str[n - i - 1];
        str[n - i - 1] = temp;
    }
}

void bitshift_string(char *str, int shift) {
    for (int i = 0; str[i] != '\0'; i++) {
        str[i] = str[i] << shift;
    }
}

void to_lower_string(char *str) {
    for (int i = 0; str[i] != '\0'; i++) {
        str[i] = tolower(str[i]);
    }
}

void to_upper_string(char *str) {
    for (int i = 0; str[i] != '\0'; i++) {
        str[i] = toupper(str[i]);
    }
}

void caesar_cipher(char *str, int shift) {
    for (int i = 0; str[i] != '\0'; i++) {
        char ch = str[i];
        if (ch >= 'a' && ch <= 'z') {
            ch = (ch - 'a' + shift) % 26 + 'a';
        } else if (ch >= 'A' && ch <= 'Z') {
            ch = (ch - 'A' + shift) % 26 + 'A';
        }
        str[i] = ch;
    }
}

File: R/string_manipulation/src/string_manipulation.rs
pub fn reverse_string(input: &str) -> String {
    input.chars().rev().collect()
}

pub fn bitshift_string(input: &str, shift: u32) -> String {
    input.chars().map(|c| (c as u8).wrapping_add(shift as u8) as char).collect()
}

pub fn to_lower_string(input: &str) -> String {
    input.to_lowercase()
}

pub fn to_upper_string(input: &str) -> String {
    input.to_uppercase()
}

pub fn caesar_cipher(input: &str, shift: u8) -> String {
    input.chars()
        .map(|c| {
            if c.is_ascii_lowercase() {
                (((c as u8 - b'a' + shift) % 26) + b'a') as char
            } else if c.is_ascii_uppercase() {
                (((c as u8 - b'A' + shift) % 26) + b'A') as char
            } else {
                c
            }
        })
        .collect()
}
